server/controllers/images.ts

```diff
@@ -0,0 +1,141 @@
+  } else {
+    return null;
+  }
+};
+
+const recordUploadedFile = async (
+  req,
+  kind,
+  file,
+  fileName,
+  s3Data: ManagedUpload.SendData | { Location: string; Bucket: string; Key: string },
+): Promise<void> => {
+  try {
+    await models.UploadedFile.create({
+      kind: kind,
+      fileName,
+      fileSize: file.size,
+      fileType: file.mimetype,
+      url: s3Data.Location,
+      data: await getUploadedFileData(file),
+      CreatedByUserId: req.remoteUser.id,
+    });
+  } catch (e) {
+    // Since this runs async, we need to report the error to Sentry in case it fails
+    console.error(e);
+    reportErrorToSentry(e, {
+      severity: 'error',
+      user: req.remoteUser,
+      transactionName: 'recordUploadedFile',
+      extra: { kind, fileName, fileType: file.mimetype },
+    });
+  }
+};
+
+const getFilename = (file, fileNameFromArgs) => {
+  const expectedExtension = SUPPORTED_FILE_EXTENSIONS[file.mimetype];
+  const rawFileName = fileNameFromArgs || file.originalname || uuid();
+  const parsedFileName = path.parse(rawFileName);
+  return `${parsedFileName.name}${expectedExtension}`;
+};
+
+export default async function uploadImage(req, res, next) {
+  const { kind } = req.body;
+
+  if (!req.remoteUser) {
+    return next(new errors.Unauthorized('You need to be logged in to upload a file'));
+  }
+
+  if (!kind) {
+    return next(
+      new errors.ValidationFailed('missing_required', {
+        kind: 'Kind field is required and missing',
+      }),
+    );
+  }
+
+  if (!SUPPORTED_FILE_KINDS.includes(kind)) {
+    const message = `Kind should be one of: ${SUPPORTED_FILE_KINDS.join(', ')}`;
+    return next(new errors.ValidationFailed('INVALID_FILE_KIND', { kind: message }, message));
+  }
+
+  // Required fields
+  const file = req.file;
+  if (!file) {
+    return next(
+      new errors.ValidationFailed('missing_required', {
+        file: 'File field is required and missing',
+      }),
+    );
+  }
+
+  // Validate file
+  if (!SUPPORTED_FILE_TYPES.includes(file.mimetype)) {
+    const message = `Mimetype of the file should be one of: ${SUPPORTED_FILE_TYPES.join(', ')}`;
+    return next(new errors.ValidationFailed('INVALID_FILE_MIME_TYPE', { file: message }, message));
+  }
+
+  if (file.size > 1024 * 1024 * 10) {
+    const message = 'Filesize cannot exceed 10MB';
+    return next(
+      new errors.ValidationFailed('INVALID_FILE_SIZE_TOO_BIG', { file: message }, message, {
+        fileSize: file.size,
+        max: '10MB',
+      }),
+    );
+  }
+
+  if (!s3) {
+    return next(new errors.ServerError('S3 service object not initialized'));
+  }
+
+  /**
+   * We will replace the name to avoid collisions
+   */
+  const fileName = getFilename(file, req.body.fileName);
+  const uploadParams = {
+    Bucket: config.aws.s3.bucket,
+    Key: `${kebabCase(kind)}/${uuid()}/${fileName || uuid()}`,
+    Body: file.buffer,
+    ACL: 'public-read',
+    ContentLength: file.size,
+    ContentType: file.mimetype,
```

Comment:
ContentDisposition for setting the download filename when accessing the file directly through s3.
Metadata to help identify the source of the file


```suggestion
    ContentType: file.mimetype,
    ContentDisposition: `attachment; filename="${fileName}"`,
    Metadata: {
      CreatedByUserId: `${req.remoteUser.id}`,
      FileKind: kind,
    },
```

Reply:
+1 for the `Metadata` field, that's useful.

Not so convinced by the `Content-Disposition` one, especially since we'll have to sanitize the `"` in the filename to make it work. I tend to prefer the `/:uuid/:filename` approach for this reason.

Reply:
If the file is saved with the original file name, I think there is no need for the content-disposition to be set.